src/circuit/hash.rs

```diff
@@ -38,10 +37,12 @@ where
             .filter(|&n| n != self.root())
         {
             let hash = hash_node(self, node, &mut hash_state);
-            hash_state.set_node(self, node, hash);
+            hash_state.set_hash(node, hash);
```

Comment:
Can I make a suggestion here? Just rename `hash_state` to `node_hashes`. Clarify that there is no state there *apart from* the node hashes, which is critical in ensuring the behaviour of the hash function is not affected by e.g. the particular topsort order that we get